mittab/apps/tab/archive.py

```diff
@@ -0,0 +1,153 @@
+from xml.etree.ElementTree import Element, SubElement, tostring
+
+from mittab.apps.tab.models import School, Team, Judge, Room, Round
+
+DEBATE_ID_PREFIX = "D"
+ROOM_ID_PREFIX = "V"
+SPEAKER_ID_PREFIX = "S"
+TEAM_ID_PREFIX = "T"
+JUDGE_ID_PREFIX = "A"
+SCHOOL_ID_PREFIX = "I"
+SPEAKER_STATUS_PREFIX = "SC"
+
+
+class ArchiveExporter:
+
+    def __init__(self, request, form):
+        self.request = request
+        self.form = form
+
+        self.data = form.cleaned_data
+        self.root = None
+
+    def export_tournament(self):
+        self.root = Element("tournament", {"name": self.data["name"]})
+        if self.data.get("host") is not None:
+            self.root.set("host", self.data["host"])
+        if self.data.get("start") is not None:
+            self.root.set("start-date", str(self.data["start"]))
+
+        self.add_rounds()
+        self.add_participants()
+        self.add_schools()
+        self.add_rooms()
+        self.add_categories()
+
+        return tostring(self.root)
+
+    def add_rounds(self):
+        pos = ["pm", "lo", "mg", "mo"]
+        qs = Round.objects.all().order_by("round_number").prefetch_related(
+            "judges", "roundstats_set")
+
+        cur_round = None
+        r_tag = None
+        for debate in qs:
+            if debate.round_number != cur_round:
+                r_tag = SubElement(self.root, "round", {
+                    "name": "Round %s" % debate.round_number
+                })
+                cur_round = debate.round_number
+
+            adjs = " ".join(
+                [JUDGE_ID_PREFIX + str(j.id) for j in debate.judges.all()])
+            d_tag = SubElement(r_tag, "debate", {
+                "id": DEBATE_ID_PREFIX + str(debate.id),
+                "chair": JUDGE_ID_PREFIX + str(debate.chair_id),
+                "adjudicators": adjs,
+                "venue": ROOM_ID_PREFIX + str(debate.room_id)
+            })
+
+            gov_tag = SubElement(d_tag, "side", {
+                "team": TEAM_ID_PREFIX + str(debate.gov_team_id)
+            })
+            opp_tag = SubElement(d_tag, "side", {
+                "team": TEAM_ID_PREFIX + str(debate.opp_team_id)
+            })
+
+            if debate.victor == Round.GOV_VIA_FORFEIT or \
+                    debate.victor == Round.ALL_DROP:
+                opp_tag.set("forfeit", "True")
+            elif debate.victor == Round.OPP_VIA_FORFEIT or \
+                    debate.victor == Round.ALL_DROP:
+                gov_tag.set("forfeit", "True")
+
+            team_points = [0, 0] # Gov, Opp
+            stats = sorted(list(debate.roundstats_set.all().values()),
+                           key=lambda x: pos.index(x["debater_role"]))
+            for i, speech in enumerate(stats):
+                side_tag = gov_tag if i % 2 == 0 else opp_tag
+                team_points[i % 2] += speech["speaks"]
```

Comment:
Not sure what the expected behavior is here, but this may not behave properly in the case of iron-people. This code is pretty messy, but you may want to look in here for some context on how scoring is done: https://github.com/MIT-Tab/mit-tab/blob/master/mittab/libs/tab_logic/stats.py

This code will work 99% of the time, but I'm not certain about 100%. Might be worth refactoring the tab_logic.stats interface to be something more useful, and then using those functions here. Specifically, functions with these signatures would be great:

```
def debater_speaks_for_round(debater, round):
	pass

def team_speaks_for_round(team, round):
	pass
```

Reply:
This is to create a "team score" by summing the speaks of the members in the round.

From what I saw, iron speakers have `RoundStats` objects for each of their positions, which is then averaged for the speaker standings. These two objects may have different values for speaks, of which should be captured in the total team speaks here. As there would still be `RoundStats` objects for each position in the case of iron-people, this would work.

A `debater_speaks_for_round(debater, round_number)` function would be inappropriate as a consequence, as that couldn't correctly handle iron-people.

Am I mistaken?

Reply:
I think it would be able to handle iron-men, but the tab logic code is all very messy. Let's leave it like this for now, and I'll dig into potentially using tab_logic some other time